### DISCUSSION TITLE
Setting database for performance

### DISCUSSION THREAD

**Discussion by cezar-radan** [codenotary/immudb#1422] [Q&A]
Hi 
I performed some load tests over an immudb database and I didn't get good scores, therefore I decided to ask for guidance related to the database settings / immudb service settings, in case of heavy load.

Testing environment: 
  test machine 
          CPU - Intel® Core™ i7-10750H CPU @ 2.60GHz × 12
          MEM - 32,0 GiB
          SSD disk 
  immudb (v 1.3.2)  - docker container running locally, on test machine
  immuclient (v 1.3.2) - used in a golang application running locally, on test machine

Test scenario is: 
Step1 - create some entries (
          - create client + open session
          - prepare 10 key-values
          - insert key-values using ExecAll (precondition key must not exist)
)
Step2 - update entries created in Step1 (
          - create client + open session
          - prepare new values for the 10 keys
          - update keys using ExecAll (precondition key must exist))
)
Step3 - get entries created in Step2(
           - create client + open session
          -  retrieve key-values using GetAll
)
          
Test Results:
* Serial running, 1 process that cycle (Step1 -> Step2 -> Step3) 1000 times.
  In this case the runtime is acceptable, example: 
                    - create client + open session = 50 - 62 milliseconds
                    - ExecAll = 10 - 13 milliseconds
                    - GetAll = 5 - 10  milliseconds
                        
* Concurrent running, 100 process that cycle (Step1 -> Step2 -> Step3) 1000 times.
  In this case the runtime is not so good, example: 
                    - create client + open session = 600 - 800 milliseconds
                    - ExecAll = 6500 - 9000 milliseconds
                    - GetAll = 100 - 150 milliseconds

So, it is very clear that in a concurrent environment MY immudb doesn't respond as expected.

I had a look over your mentions and over your code :
        https://github.com/codenotary/immudb#performance-figures   
        https://github.com/codenotary/immudb/tree/master/tools/testing/stress_tool_test_kv
         https://github.com/codenotary/immudb-tools/tree/main/stresstest
and, it seems there are various database parameters/configs that could be changed in order to improve the performance:
maxConcurrency (30 <-> 100)
maxIOConcurrency (1 <-> 100)
indexSettings.cleanupPercentage (0.1 <-> 10)
.... so on
I have run tests with various database parameters values (many combinations) but, unfortunately, I didn't obtain any improvement. 

Please, could you provide some guidance for database settings/configs in order to obtain a better performance (heavy inserts) ?
I am pretty sure, in my case, the indexing process should be configured with care too.

Thank you very much.

**Comment by jeroiraz** [comment]
Hi @cezar-radan, thanks for reaching out. Your description is very well described :)

Currently, to get benefits from concurrent writers, transactions must be read-only e.g. Set, SetAll.

ExecAll is acquiring a lock, thus why increasing the number of writers is not producing a significative benefit.

Multi-version Concurrency Control (MVCC) is partially implemented but will be fully implemented in the near future. With that, it will be possible to obtain benefits from concurrency when the transaction does not only involves writes but read and writes.

**Reply by cezar-radan** [reply]
Hi @jeroiraz 
I am looking forward for the MVCC version.

One more question, please.
During load tests, described above, many times the immudb system became unresponsive. It didn't allow to open or to create new connections.
For example, when immudb becomes unresponsive:
- immuadmin / immuclient cannot connect to immudb, those clients run forever since cannot acquire a connection to the blocked immudb.  
-  but, immudb's metrics endpoint (http://localhost:9497/metrics) provides a response, if I call it.

Of course, I need to restart immudb container in order to become usable.

What I have noticed into the immudb log ->
1. There is no error registered into the log
2. When immudb becomes unresponsive, last records into the log are:
> INFO: Flushing index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30, since_cleanup=48750} requested via FlushWith...
>  INFO: Index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30} successfully flushed
>  INFO: Index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30} successfully synced
>  INFO: Discarding unreferenced data at index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30, current_min_offset=28965262487, new_min_offset=29232904420}...
>  INFO: Unreferenced data at index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30, current_min_offset=28965262487, new_min_offset=29232904420} successfully discarded
>  INFO: Discarding older snapshots at index '/var/lib/immudb/test1/index' {ts=1895818, opened_snapshots=1}...
>  INFO: Older snapshots at index '/var/lib/immudb/test1/index' {ts=1895818, opened_snapshots=1} successfully discarded
>  INFO: Flushing index '/var/lib/immudb/test1/index' {ts=1895818} finished with: 36 inner nodes, 75 leaf nodes, 628 entries
> DEBUG: Open sessions count: 2
> DEBUG: Inactive sessions count: 0
> DEBUG: Deleted sessions count: 0

3. After immudb restart (restart is performed after 2h since immudb becomes unresponsive):
> INFO: Syncing Binary Linking at '/var/lib/immudb/test1'...
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1801000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1802000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1803000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1804000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1805000
> ........................... so on............................................
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1891000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1892000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1893000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1894000
> INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1895000
> INFO: Binary Linking up to date at '/var/lib/immudb/test1'
> INFO: Reading snapshots at '/var/lib/immudb/test1/index/commit'...
> INFO: Index '/var/lib/immudb/test1/index' {ts=1895818, discarded_snapshots=0} successfully loaded
> INFO: Successfully read snapshots at '/var/lib/immudb/test1/index/commit'
> INFO: Indexing in progress at '/var/lib/immudb/test1'
> INFO: Database 'test1' {replica = false} successfully opened

What should I change in order to stop getting immudb system stuck ? Maybe there are specific immudb settings that should be accommodated for load tests (change default values) or OS settings( ex. max open files per process) that should be changed ?

Thank you very much.

**Reply by jeroiraz** [reply]
> Hi @jeroiraz I am looking forward for the MVCC version.
> 
> One more question, please. During load tests, described above, many times the immudb system became unresponsive. It didn't allow to open or to create new connections. For example, when immudb becomes unresponsive:
> 
> * immuadmin / immuclient cannot connect to immudb, those clients run forever since cannot acquire a connection to the blocked immudb.
> * but, immudb's metrics endpoint (http://localhost:9497/metrics) provides a response, if I call it.
> 
> Of course, I need to restart immudb container in order to become usable.
> 
> What I have noticed into the immudb log ->
> 
> 1. There is no error registered into the log
> 2. When immudb becomes unresponsive, last records into the log are:
> 
> > INFO: Flushing index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30, since_cleanup=48750} requested via FlushWith...
> > INFO: Index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30} successfully flushed
> > INFO: Index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30} successfully synced
> > INFO: Discarding unreferenced data at index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30, current_min_offset=28965262487, new_min_offset=29232904420}...
> > INFO: Unreferenced data at index '/var/lib/immudb/test1/index' {ts=1895818, cleanup_percentage=0.30/0.30, current_min_offset=28965262487, new_min_offset=29232904420} successfully discarded
> > INFO: Discarding older snapshots at index '/var/lib/immudb/test1/index' {ts=1895818, opened_snapshots=1}...
> > INFO: Older snapshots at index '/var/lib/immudb/test1/index' {ts=1895818, opened_snapshots=1} successfully discarded
> > INFO: Flushing index '/var/lib/immudb/test1/index' {ts=1895818} finished with: 36 inner nodes, 75 leaf nodes, 628 entries
> > DEBUG: Open sessions count: 2
> > DEBUG: Inactive sessions count: 0
> > DEBUG: Deleted sessions count: 0
> 
> 3. After immudb restart (restart is performed after 2h since immudb becomes unresponsive):
> 
> > INFO: Syncing Binary Linking at '/var/lib/immudb/test1'...
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1801000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1802000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1803000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1804000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1805000
> > ........................... so on............................................
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1891000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1892000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1893000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1894000
> > INFO: Binary linking at '/var/lib/immudb/test1' in progress: processing tx: 1895000
> > INFO: Binary Linking up to date at '/var/lib/immudb/test1'
> > INFO: Reading snapshots at '/var/lib/immudb/test1/index/commit'...
> > INFO: Index '/var/lib/immudb/test1/index' {ts=1895818, discarded_snapshots=0} successfully loaded
> > INFO: Successfully read snapshots at '/var/lib/immudb/test1/index/commit'
> > INFO: Indexing in progress at '/var/lib/immudb/test1'
> > INFO: Database 'test1' {replica = false} successfully opened
> 
> What should I change in order to stop getting immudb system stuck ? Maybe there are specific immudb settings that should be accommodated for load tests (change default values) or OS settings( ex. max open files per process) that should be changed ?
> 
> Thank you very much.

hm, this seems to be an issue. Could you provide detailed information on the steps to reproduce it?

**Reply by cezar-radan** [reply]
Sure
The test is very similar with the one described above, only the test machine was changed, but I will resume.

Scenario:
Step1 - create some entries (
- prepare 10 key-values
- insert key-values using SetAll (precondition key must not exist)
)
Step2 - update entries created in Step1 (
- prepare new values for the 10 keys
- update keys using SetAll (precondition key must exist))
)
Step3 - get entries updated in Step2(
- retrieve key-values using GetAll
)

Test executed:
100 concurrent processes (goroutines). Each process (gorounite) executes Step1 -> Step2 -> Step3, 1000 times. 

Test environment:
Machine
OS macOS Monterey version 12.5.1
Processor 2,3 GHz 8-Core Intel Core i9
Memory 16GB 2667 MHz DDr4
SSD disk
 
Immudb(v 1.3.2) - docker container running locally, on test machine

Database - since the test was executed in a concurrent environment I have set (MaxConcurrency:100, 	MaxIOConcurrency:10)

**Reply by cezar-radan** [reply]
Update.
It seems that immudb unresponsiveness is due to database's parameters "MaxConcurrency:100" and "MaxIOConcurrency:10".

I repeated the test (couple times) using default values for "MaxConcurrency:30" and "MaxIOConcurrency:1", and immudb was still functional at the end of the test.
Of course, as normally, using default parameter's values, the test execution time was longer.

**Reply by byo** [reply]
Hi @cezar-radan , the slowdown may be due to preconditions used in Steps 2 and 3 - can you do a test where those preconditions are not present just to check if there's a performance difference?

**Reply by cezar-radan** [reply]
Hi @byo. I removed the Step2 and Step3 preconditions but, unfortunately, I didn't get any significant runtime improvement.

**Reply by cezar-radan** [reply]
Update - What I have noticed.
My test uses concurrent goroutines (100), and each goroutine manages itself the connection to immudb (create client -> open session -> close session).  
I have changed a little bit and I have created a single/unique client+session. 
If all goroutines use this unique connection (client + session) the runtime decreases by 3 times, and this is a tremendous improvement.

**Reply by byo** [reply]
@cezar-radan that's a very interesting finding, what was the flow of OpenSession / CloseSession before switching to a single client? Was the session opened / closed for a single round of changes? Even with multiple sessions it should work faster assuming that the session is opened once, reused between rounds and only closed before the test is finished. If it is not working like that then it is definitely something to investigate on immudb side.

**Reply by cezar-radan** [reply]
Before the unique client+session, each goroutine opens (once) a client+session, do all the work and, at the end, close the session. 
So, 100 concurrent workers (goroutines) means 100 connections (client+session) open at the same time.

**Reply by cezar-radan** [reply]
I have tried, also, the version where a goroutine deals with multiple connections. This means for each round/step a goroutine performs create client + open session + close session, but the runtime was not really encouraging.
After that, I have used a single connection for each goroutine and I have noticed a good improvement.
This determine me to create a unique connection (client+session) and to use it for all goroutines.
